include/turnstile.h

```diff
@@ -10,7 +10,8 @@ void turnstile_init(turnstile_t *);
 
 /* This puts currently running thread on the turnstile,
  * puts it in waiting state and yields. While waiting on turnstile
- * thread cannot be run. This function is done under critical section. */
+ * thread cannot be run. This function has to be called with critical section,
```

Comment:
I think I can justify why they do it like this. When you have SMP architecture disabling interrupts isn't enough (other processor might still change the value), but blocking spin lock is enough. So the question is where to put the spin lock. Obviously we should lock the spin lock when trying to modify the turnstile, but the spinlock is inside the turnstile itself so should we exctract it and lock it outside, or just embed it in function and change invariants. Note that since FreeBSD uses hashing on turnstile they have more locks to lock than just one. Also similar situations with race condition and missed wake-ups might occur with implementation of other types of mutexes, these weird invariants might be a hint that one should remember about possible race conditions.